integration-tests/src/tests/client/features/in_memory_tries.rs

```diff
@@ -0,0 +1,361 @@
+            account: Account::new(initial_balance, staked, CryptoHash::default(), 0),
+        });
+        records.push(StateRecord::AccessKey {
+            account_id: account.clone(),
+            public_key: create_user_test_signer(&account).public_key,
+            access_key: AccessKey::full_access(),
+        });
+        // The total supply must be correct to pass validation.
+        genesis_config.total_supply += initial_balance + staked;
+    }
+    let genesis = Genesis::new(genesis_config, GenesisRecords(records)).unwrap();
+    let chain_genesis = ChainGenesis::new(&genesis);
+
+    // Create two stores, one for each node. We'll be reusing the stores later
+    // to emulate node restarts.
+    let stores = vec![create_test_store(), create_test_store()];
+    let mut env = TestEnv::builder(chain_genesis.clone())
+        .clients(vec!["account0".parse().unwrap(), "account1".parse().unwrap()])
+        .stores(stores.clone())
+        .real_epoch_managers(&genesis.config)
+        .track_all_shards()
+        .nightshade_runtimes_with_trie_config(
+            &genesis,
+            vec![
+                TrieConfig::default(), // client 0 does not load in-memory tries
+                TrieConfig {
+                    // client 1 loads two of four shards into in-memory tries
+                    load_mem_tries_for_shards: vec![
+                        ShardUId { version: 1, shard_id: 0 },
+                        ShardUId { version: 1, shard_id: 2 },
+                    ],
+                    max_mem_tries_size_per_shard: 100 * 1024 * 1024,
+                    ..Default::default()
+                },
+            ],
+        )
+        .build();
+
+    // Sanity check that we should have two block producers.
+    assert_eq!(
+        env.clients[0]
+            .epoch_manager
+            .get_epoch_block_producers_ordered(
+                &EpochId::default(),
+                &env.clients[0].chain.head().unwrap().last_block_hash
+            )
+            .unwrap()
+            .len(),
+        2
+    );
+
+    // First, start up the nodes from genesis. This ensures that in-memory
+    // tries works correctly when starting up an empty node for the first time.
+    let mut nonces =
+        accounts.iter().map(|account| (account.clone(), 0)).collect::<HashMap<AccountId, u64>>();
+    let mut balances = accounts
+        .iter()
+        .map(|account| (account.clone(), initial_balance))
+        .collect::<HashMap<AccountId, u128>>();
+
+    run_chain_for_some_blocks_while_sending_money_around(&mut env, &mut nonces, &mut balances, 100);
+    // Sanity check that in-memory tries are loaded, and garbage collected properly.
+    // We should have 1 unique root because the last few blocks should all have the
+    // same state roots.
+    assert_eq!(MEM_TRIE_NUM_ROOTS.get_metric_with_label_values(&["s0.v1"]).unwrap().get(), 1);
+    assert_eq!(MEM_TRIE_NUM_ROOTS.get_metric_with_label_values(&["s2.v1"]).unwrap().get(), 1);
```

Comment:
Can we pull here `MemTries` with read-only access and query `roots` directly? I would avoid relying on prometheus module, especially if we want to modify set of metrics at some point. Also I remember an issue - if two tests rely on the same prometheus metric, it will be spoiled because of concurrent access.

It applies to `test_trie_recording` as well.

Reply:
Ah... I see. Flakiness due to concurrent access. Good point.

Reply:
For trie_recording tests, it should be less of an issue because I'm only asserting that the metric goes up. And that metric only goes up, never down.